### ISSUE TITLE
Dataloader stuck with AseEnvironmentProvider

### ISSUE THREAD

**Issue by Dom1L** [atomistic-machine-learning/schnetpack#96]
Hi guys,

thanks for the suggestions with the AseEnvironmentProvider. 
It worked well with my dummy test set of 100 structures, but I discovered that for a larger database size the whole data loading procedure gets stuck. This means when the data is initialized the first time (either during `loader.get_statistics()` or in the trainer function later) it doesn't finish. I also had an issue where it continued to fill up my ram and I needed to stop at 90Gb. 
I set the number of workers in the dataloader to 0 to debug a bit and it seems to be stuck in the ase neighbour list function.
I tested this also with small molecules and its reproducible with the qm9 dataset.

Have you experienced something similar? 

I used the same script as before with the AseEnvironment addition. The issue already arises with a database size of 3000 molecules.

Thanks for your help!

```python
import torch
import torch.nn.functional as F
from torch.optim import Adam

import schnetpack as spk
from schnetpack.data import AtomsData
import schnetpack.atomistic as atm
import schnetpack.representation as rep

cutoff = 5.  # Angstrom
environment_provider = spk.environment.AseEnvironmentProvider(cutoff)
data = AtomsData('trypsin.db', required_properties=['energy'], environment_provider=environment_provider)

# split in train and val
train, val, test = data.create_splits(2800, 100)
loader = spk.data.AtomsLoader(train, batch_size=10, num_workers=4)
val_loader = spk.data.AtomsLoader(val)

# create model
reps = rep.SchNet(
    n_atom_basis=128,
    n_filters=128,
    n_interactions=1,
    cutoff=5.0,
    n_gaussians=25,
    normalize_filter=False,
    coupled_interactions=False,
    return_intermediate=False,
    max_z=100,
    trainable_gaussians=False,
    distance_expansion=None)
output = atm.Atomwise()
model = atm.AtomisticModel(reps, output).cuda()

opt = Adam(model.parameters(), lr=1e-3)
loss = lambda b, p: F.mse_loss(p["y"], b['energy'])
out_dir = 'test'
trainer = spk.train.Trainer(out_dir, model, loss, opt, loader, val_loader,
                            hooks=[spk.train.CSVHook(f'{out_dir}/log',
                                                     [spk.metrics.MeanAbsoluteError('energy', "y"),
                                                      spk.metrics.RootMeanSquaredError('energy', "y")],
                                                     every_n_epochs=1)])
trainer.train(torch.device("cuda"))
```

**Comment by Dom1L** [comment]
Hi,

I was able to make this issue occur with an even smaller dataset. I attached an example that contains 100 binding pockets and some dummy energies which in combination with the script above fills up the memory entirely.
[test.zip](https://github.com/atomistic-machine-learning/schnetpack/files/3063590/test.zip)

So to clarify a bit: 
* Many atoms + AseEnvironmentProvider -> memory (+ loading) issue
* Small molecules + AseEnvironmentProvider -> loading issue

I have found an issue on the pytorch repository related to memory leaks with their Dataloader (https://github.com/pytorch/pytorch/issues/13246), I don't know whether this is related to this issue or not since the memory issue also occurs for num_workers=0, but I thought it might be helpful.

I also tested different pytorch versions (0.41 and 1.0.1.post2) and I can reproduce this with both.

Thanks a lot!

**Comment by Stefaanhess** [comment]
Thanks, for the additional information. I will have a look at it!

**Comment by Dom1L** [comment]
Hey here some updates,

so after some further investigations I was able to pinpoint the memory issues etc. to some specific matrix operations in the ASE neighbourlist function. So not a memory leak issue with the Dataloader which is good news.
What I then did was to write a new environment provider that utilizes neighbourlist code from the Torchani repository, which made it able to train on batches of size 16 with ~1000 atoms. The downside is that it is fairly slow for now, so there is some room for improvement. It definitely made it work for now, but I haven't tested it thoroughly so far

**Comment by ktschuett** [comment]
Ok, so it is a bug in ASE? In their docs they write:
"Atoms outside nonperiodic boundaries are included in the neighbor list but complexity of neighbor list search for those can become n^2."

Could it be that if the cell is not set large enough to contain the molecule, this also causes a memory leak for some reason?

**Comment by Dom1L** [comment]
Could be... I think that none of my samples have the cell explicitly set which might cause problems in ASE. 
Which would also explain why my trypsin example is working because I read it from the pdb directly, the other stuff is preprocessed etc.
I'll check and report later, thanks for the tip!

**Comment by Dom1L** [comment]
@ktschuett you were right with your suspicion. I preprocessed the molecules with HTMD and then once I convert them into ASE atoms they have a non defined cell and pbc=False. For the trypsin case, I read it directly from the pdb using ASE and all properties were automatically defined. So the reason I could observe the same behaviour for the QM9 data set is that these molecules also don't have defined cells. So might be worth it to implement a check or something in case someone uses that option I don't know...

So in the end with the cells correctly defined it worked using the AseEnvironmentProvider, but was roughly 3x slower than my TorchAniEnvironmentProvider( ~144s/epoch with ASE, ~47s/epoch TorchAni on 3k samples on a TitanV).

So if you're interested I could help out a bit with that.
Thanks for the help!

**Comment by ktschuett** [comment]
Yes, if you would like to provide a pull request for this, that would be great!
